sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -333,6 +335,39 @@ object ProjectCollapsing extends Rule[LogicalPlan] {
         )
         Project(cleanedProjection, child)
       }
+
+    // TODO Eliminate duplicate code
+    // This clause is identical to the one above except that the inner operator is an `Aggregate`
+    // rather than a `Project`.
+    case p @ Project(projectList1, agg @ Aggregate(_, projectList2, child)) =>
```

Comment:
If we don't do this, then we will generate something like:
`select i + 1 from (select sum(b) as i from tbl group by a) t`.

Actually it's same with `select sum(b) + 1 from tbl group by a`, although these 2 sql string doesn't equal. I think our goal is to generate "semantically right" sql string,  not "exactly same" sql string. Please correct me if I missed something here.